compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -588,3 +588,20 @@ pub fn read_target_uint(endianness: Endian, mut source: &[u8]) -> Result<u128, i
     debug_assert!(source.len() == 0); // We should have consumed the source buffer.
     uint
 }
+
+/// Read a pointer-sized thing from a target, so it will fit in a u64
+/// as the modules using this don't support wider pointer values
+/// FIXME(jubilee): Move this out of here and closer to the things using it!
+#[inline]
+pub fn read_target_ptr(endian: Endian, ptr_size: usize, bytes: &[u8]) -> u64 {
+    use core::convert::TryInto;
+    match (ptr_size, endian) {
+        (2, Endian::Little) => u16::from_le_bytes(bytes.try_into().unwrap()) as u64,
+        (2, Endian::Big) => u16::from_be_bytes(bytes.try_into().unwrap()) as u64,
+        (4, Endian::Little) => u32::from_le_bytes(bytes.try_into().unwrap()) as u64,
+        (4, Endian::Big) => u32::from_be_bytes(bytes.try_into().unwrap()) as u64,
+        (8, Endian::Little) => u64::from_le_bytes(bytes.try_into().unwrap()),
+        (8, Endian::Big) => u64::from_be_bytes(bytes.try_into().unwrap()),
```

Comment:
I don't like explicitly handling several possibilities using host integers, when it can be done by e.g. copying `bytes` into an `[0u8; 8]`, reversing the `..bytes.len()` slice of the array if `Endian::Big`, and then using `u64::fom_le_bytes`.

@oli does miri already have some uniform (over integer sizes) code for doing this?

Actually, why not just keep using `read_target_uint`? It seems like the correct abstraction, am I missing something? If you don't want it to part of the `rustc_mir::mir::interpret` interface, we just have to move it around a bit, right?

Reply:
> does miri already have some uniform (over integer sizes) code for doing this?

Yes, as you noted, read_target_uint is that abstraction. I do see the point in an abstraction returning u64 for usize purposes, but imo that should just be a wrapper around the existing functions.

As to where to place them... if rustc_middle doesn't need them they can go to rustc_mir, otherwise idk

Reply:
Hmm, I definitely was overthinking this, in retrospect. Anyways, my main thought was that it should probably exist in, say, `rustc_target` or something like that.